feat: add support for refreshKeys

kissmikijr opened the pull request:
## Hey, I just made a Pull Request!
 This is a potential implementation of the https://github.com/backstage/backstage/issues/12155 Nothing set in stone, would like some feedback on it!

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

kissmikijr commented on plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:
```diff
@@ -515,7 +518,40 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
+    const tx = txOpaque as Knex.Transaction;
+    const { keys } = options;
+
+    await Promise.all(
+      keys.map(k => {
+        return tx<DbRefreshKeysRow>('refresh_keys')
+          .insert({
+            entity_ref: stringifyEntityRef(k.entity),
+            key: k.key,
+          })
+          .onConflict(['entity_ref', 'key'])
+          .ignore();
```
I wasn't able to perform this with tx.batchInsert when I want to use an onConflict. Would it be better to check first if they exist and only do an insert if don't?

kissmikijr commented on plugins/catalog-backend/migrations/20220616202842_refresh_keys.js:
```diff
@@ -0,0 +1,41 @@
+ */
+
+exports.up = async function up(knex) {
+  await knex.schema.createTable('refresh_keys', table => {
```
Not sure if this table should have a dedicated id field ?

kissmikijr commented:
> Not sure if I should comment on the RFC or the PRFC might be good to combine them next time 😅
Yeah I agree, I'll definietly combine them next time, it is a littlebit awkward this way. :/

kissmikijr commented on plugins/catalog-backend/src/database/types.ts:
```diff
@@ -81,6 +81,14 @@ export type ReplaceUnprocessedEntitiesOptions =
     };
 
+export type RefreshKeyOptions = {
+  keys: { key: String; entity: Entity }[];
```
Could be, however in this case it would mean we'd need to emit the entityRefs from the processors, so the entityRef construction would happen inside all of the processors. With the current implementation they just emit an entity, and in the processing engine we can construct the refs based on the emitted entity. 

I'm fine with both of them, maybe leaning towards the emission of entityRefs just for the sake of reducing the data thatis sent :D

kissmikijr commented:
@jhaals With this my intention was to understand the problem space and try to explore the way it can be done, if you think in a general sense it looks okey, I'd move forward with adding tests, and making sure it passes the ci and move it to a regular PR.

jhaals commented on plugins/catalog-backend/src/service/DefaultRefreshService.ts:
```diff
@@ -45,4 +49,9 @@ export class DefaultRefreshService implements RefreshService {
       });
     });
   }
+  async refreshByRefreshKey(options: RefreshByRefreshKeysOptions) {
```
Worth renaming this method and make it accept multiple keys.

```suggestion
  async refreshByRefreshKeys(options: RefreshByRefreshKeysOptions) {
```

jhaals commented on plugins/catalog-backend/api-report.md:
```diff
@@ -601,6 +613,7 @@ export const processingResult: Readonly<{
     newEntity: Entity,
   ) => CatalogProcessorResult;
   readonly relation: (spec: EntityRelationSpec) => CatalogProcessorResult;
+  readonly refresh: (entityRef: string, key: string) => CatalogProcessorResult;
```
As far as we can tell at this point we only need to refresh emit refresh keys for the entity that is currently being processed. I suggest we drop the entityRef from the result and leave it to be added by the `updateProcessedEntity` method. In that case the `setRefreshKeys` method is no longer needed on the database.

kissmikijr replied:
Yea, makes sense and much nicer imo!

jhaals commented on plugins/catalog-backend/src/modules/core/UrlReaderProcessor.ts:
```diff
@@ -83,6 +83,14 @@ export class UrlReaderProcessor implements CatalogProcessor {
+          if (parseResult.type === 'entity') {
+            emit(
+              processingResult.refresh(
+                stringifyEntityRef(parseResult.entity),
```
Same applies here as to the file reader processor :)

jhaals commented on plugins/catalog-backend/src/modules/core/UrlReaderProcessor.ts:
```diff
@@ -83,6 +83,14 @@ export class UrlReaderProcessor implements CatalogProcessor {
+            emit(
+              processingResult.refresh(
+                stringifyEntityRef(parseResult.entity),
+                item.url,
```
One thing we realised reviewing is that glob pattern locations won't support immediate refreshes of new files that are added since the glob pattern is expanded only to files that already exist. Do you have any ideas on how we could express glob patterns as refresh keys?

We where thinking that we might end up with the possibility of emitting a pattern like `https://example.com/foo/*` and matching that either in the database or a separate processor method, but we're open to ideas.

Probably let's skip it for this PR and leave it as an unknown problem for the future.

jhaals commented on plugins/catalog-backend/migrations/20220616202842_refresh_keys.js:
```diff
@@ -0,0 +1,41 @@
+      'This table contains relations between entities and keys to trigger refreshes with',
+    );
+    table
+      .text('entity_ref')
```
Let's make the foreign key constraint on the entity_id in the refresh_state table and also add a cascading delete so that we clear items in this table when we drop from refresh_state.

See https://github.com/backstage/backstage/blob/7187adfadd7d06498e015f3ac26eaaf5ebc8c01d/plugins/catalog-backend/migrations/20210302150147_refresh_state.js#L152 for inspiration

Rugvip replied:
@kissmikijr did you tun into trouble using and referencing `entity_id` here instead? It's the PK so a bit nicer to use that rather than the ref. I believe it should be available when needed as well.

kissmikijr replied:
Nope, I actually thought we will need the entity_ref, but just realized that the id is more than enough since we are going to refresh based on the key anyways.

jhaals commented on plugins/catalog-backend/migrations/20220616202842_refresh_keys.js:
```diff
@@ -0,0 +1,41 @@
+      .comment(
+        'A reference to a key which should be used to trigger a refresh on this entity',
+      );
+    table.unique(['entity_ref', 'key']);
```
We should delete this line and replace with an index like we do here 
https://github.com/backstage/backstage/blob/7187adfadd7d06498e015f3ac26eaaf5ebc8c01d/plugins/catalog-backend/migrations/20210302150147_refresh_state.js#L168

We also need one index for the key column since we are going to find by key.

kissmikijr replied:
If we removes this wouldnt it mean that the same record can get inserted everytime the processors runs? So possibly on every refresh ?
I wanted to avoid this by adding a unique constraint here and handling it on the insert by onconflict ignore.

Rugvip replied:
We'll also want to delete all existing entries for the given entity whenever we write a processing result, so that should be np

jhaals commented on plugins/catalog-backend/migrations/20220616202842_refresh_keys.js:
```diff
@@ -0,0 +1,41 @@
+ * @param { import("knex").Knex } knex
+ */
+exports.down = async function down(knex) {
+  await knex.schema.dropTable('refresh_keys');
```
remember to drop the newly added indices here as well with an alter table. 

https://github.com/backstage/backstage/blob/7187adfadd7d06498e015f3ac26eaaf5ebc8c01d/plugins/catalog-backend/migrations/20210302150147_refresh_state.js#L200

jhaals reviewed (commented):
Hey @kissmikijr, sorry to keep you waiting on the bigger review!
Me, @Rugvip and @freben looked provided some comments, great work!

kissmikijr commented on plugins/catalog-backend/src/modules/core/FileReaderProcessor.ts:
```diff
@@ -61,6 +62,14 @@ export class FileReaderProcessor implements CatalogProcessor {
+            if (parseResult.type === 'entity') {
+              emit(
+                processingResult.refresh(
+                  stringifyEntityRef(parseResult.entity),
```
yea, I figured this out yesterday the hard way! :D Was checking the code what is going on why it is not refreshing my entities! :D 
Thank you, I'm on it!

kissmikijr replied:
In the meantime I realized what the other comments mean, re written and works like a charm, without emitting the entity from the processors and using the one in the `updateProcessedEntity`

kissmikijr commented on plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:
```diff
@@ -100,11 +102,12 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
         `Conflicting write of processing result for ${id} with location key '${locationKey}'`,
       );
     }
+    const sourceEntityRef = stringifyEntityRef(processedEntity);
```
@Rugvip As I was testing this implementation with the different usecases I came across that when the PlaceholderProcessor runs the processedEntity will be the actual entity not the location. So if I store the entityRef of that entity in the refreshKeys table I assume the refresh won't happen because in those cases we also should save the location entity. 

I can see something like this happening in the refresh service with the listAncestors method. Which place would be the best suited to find the ancestor location of an entity ?
I think it could happen here, so we only store location entityRefs in the refreshKeys table, or it could happen at the refreshByRefreshKeys method where we would check if the entity ref is a location if not first find its ancestor location and then trigger a refresh.

Rugvip replied:
No that's all good actually, for the `PlaceholderProcessor` it is the entity itself that we need to process, not the location. I'd say avoid thinking of the processing in terms of what roles different entities have, and instead what the processors do given different input. If you want something to happen again, then you just want to run the processor again with the same kind of input.

One thing to keep in mind as well is that `readLocation` is just an abstraction around `Location` entities, so essentially `readLocation` is the same thing as processing a `Location` entity.

Rugvip commented on plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:
```diff
@@ -138,6 +141,19 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
     );
 
+    // Insert the refresh keys for the processed entity
+    await Promise.all(
+      options.refreshKeys.map(k => {
+        return tx<DbRefreshKeysRow>('refresh_keys')
+          .insert({
+            entity_ref: sourceEntityRef,
+            key: k.key,
+          })
+          .onConflict(['entity_ref', 'key'])
+          .ignore();
+      }),
+    );
```
You can handle this the same way as the relations above, but even a bit simpler because we don't care about the previous values. So delete all entries for the given entity, then batch insert all the new ones. No need to handle conflicts since they won't happen with that strategy

Rugvip commented on plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:
```diff
@@ -516,6 +532,31 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
+    const updateResult = await tx<DbRefreshStateRow>('refresh_state')
+      .whereIn('entity_ref', function selectEntityRefs(tx2) {
+        tx2
+          .whereIn('key', keys)
```
It's probably worth doing a lil bit of manual batching here tbh, chunking the keys up by however many values `min(SQLite, PostgreSQL)` supports for `IN`. Could easily have a PR that changes thousands of files and have this blow up right?

kissmikijr replied:
Based on some google searches it seems that postgres doesn't have direct limit for the IN condition, the implicit limit is 32767. For SQLite there aren't explicit limit for the IN condition either, but the maximum_length_of a query is: 1000000

In this case you still think we should do some manual batching ? It seem the in condition should be able to handle some pretty big arrays.

Rugvip commented on plugins/catalog-backend/src/database/DefaultProcessingDatabase.ts:
```diff
@@ -516,6 +532,31 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
+      .update({ next_update_at: tx.fn.now() });
+
+    if (updateResult === 0) {
+      throw new NotFoundError(
+        `Failed to schedule ${JSON.stringify(keys)} for keys`,
+      );
```
I don't think this should be an error tbh? Could easily have attempts to refresh with 0 hits

kissmikijr replied:
Correct, this shouldn't be an error.

Rugvip commented on plugins/catalog-backend/src/modules/core/FileReaderProcessor.ts:
```diff
@@ -56,11 +58,16 @@ export class FileReaderProcessor implements CatalogProcessor {
             emit(parseResult);
+            emit(
+              processingResult.refresh(
+                `${LOCATION_TYPE}:${path.normalize(fileMatch)}`,
```
```suggestion
                `${LOCATION_TYPE}://${path.normalize(fileMatch)}`,
```

Are these always absolute paths btw? I'm thinking that would likely be best

kissmikijr replied:
Based on my testing the npm glob library will return the absolute paths in this case, so here the fileMatch will always be absolute.

Rugvip commented on plugins/catalog-backend/src/modules/core/PlaceholderProcessor.ts:
```diff
@@ -127,6 +134,8 @@ export class PlaceholderProcessor implements CatalogProcessor {
           base,
         });
 
+      emit(processingResult.refresh(`url:${resolverValue}`));
```
Not actually the correct value here right? It should be the absolute resolved target URL. Looking like this would be the relative one

kissmikijr replied:
yea I think you are correct here, this will be the relative one

Rugvip commented on plugins/catalog-backend/src/modules/core/UrlReaderProcessor.ts:
```diff
@@ -93,6 +93,8 @@ export class UrlReaderProcessor implements CatalogProcessor {
         });
       }
+
+      emit(processingResult.refresh(`${location.type}:${location.target}`));
```
I'm thinking we want just the URL here rather than `<type>:<target>`, what do you think @backstage/maintainers?

Rugvip reviewed (changes requested):
Another round! 🍻

Rugvip commented on plugins/catalog-backend/src/database/types.ts:
```diff
@@ -81,6 +82,14 @@ export type ReplaceUnprocessedEntitiesOptions =
       type: 'delta';
     };
 
+export type RefreshKeyOptions = {
+  refreshKeys: RefreshKeyData[];
+};
```
Used right?

Rugvip commented on plugins/catalog-backend/src/modules/core/PlaceholderProcessor.ts:
```diff
@@ -127,6 +134,18 @@ export class PlaceholderProcessor implements CatalogProcessor {
         });
 
+      emit(
+        processingResult.refresh(
+          `url:${relativeUrl({
+            key: resolverKey,
+            value: resolverValue,
+            baseUrl: location.target,
+            read,
+            resolveUrl,
+          })}`,
+        ),
```
I think this is gonna work for a lot of `resolver` implementations, but not all. I think it's worth an investigation into what it would mean to let the `resolver` return or emit refresh keys. Think for example an OpenAPI-aware resolver that follows `$ref` URLs that we then add to the set of refresh keys

kissmikijr replied:
Agree, that's gonna be more future-proof for sure! Let me see how would it look.

Rugvip commented on plugins/catalog-backend/src/modules/core/FileReaderProcessor.ts:
```diff
@@ -50,17 +52,23 @@ export class FileReaderProcessor implements CatalogProcessor {
             emit(parseResult);
+            emit(
+              processingResult.refresh(
+                `${LOCATION_TYPE}://${normalizedFilePath}`,
```
Sorry for flip-flopping, but we're leaning back towards using the serialized location format for refresh keys after all. Feels like that's the more future proof design that will let us add new behavior without awkwardness.

Think for example a refresh key that simply references a GitHub repo, we'd probably do something like `github-repo:https://github.com/backstage/backstage` for that since we need pretty much all parts of that URL anyway. If we just use the scheme it would need to either be the bare `https://github.com/backstage/backstage`, but there's a potential risk for overlap there. Another more explicit option is then `github-repo://github.com/backstage/backstage`, but that's where it starts getting real awkward, especially of we have values that don't really fit as URLs.

So with that, let's switch this back to `file:<path>` for now, as in drop the `//` 😅

kissmikijr commented:
@Rugvip Addressed the comments, opted to add the emit to the resolvers as the easiest solution right now. I can see some tests failed, but they pass locally, can it be that it is flaky?

Rugvip commented on plugins/catalog-backend/src/service/types.ts:
```diff
@@ -71,6 +74,7 @@ export interface RefreshService {
    * Request a refresh of entities in the catalog.
    */
   refresh(options: RefreshOptions): Promise<void>;
+  refreshByRefreshKeys(options: RefreshByRefreshKeysOptions): Promise<void>;
```
What do you think about making this an addition to the existing `RefreshOptions` rather than a new method?

kissmikijr replied:
If we won't expose the refreshKeys on the DefualtRefreshService we won't need to add the options to the existing RefreshOptions either, so I'll hold on to this and do it based on how we decide.

Rugvip commented on plugins/catalog-backend/src/service/DefaultRefreshService.ts:
```diff
@@ -45,4 +49,9 @@ export class DefaultRefreshService implements RefreshService {
       });
     });
   }
+  async refreshByRefreshKeys(options: RefreshByRefreshKeysOptions) {
+    await this.database.transaction(async tx => {
+      await this.database.refreshByRefreshKeys(tx, options);
+    });
+  }
```
I feel we need to do something different here tbh, just passing through without authorization is pretty unexpected. What I'm thinking is that we'd keep refreshing by keys internal, so it's not a thing you can request via the API. The only way to refresh via keys would be some form of direct runtime access for example via the entity providers API.

Have you considered this step yet? didn't see it in the RFC. How do you end up wanting to pipe refresh data into the catalog? Maybe we could even skip this part in the PR and leave it to a followup

kissmikijr replied:
Yea, I agree my initial idea was to just expose it on the refresh service and have its own endpoint. I can accept that we do not want to do this.

The other thing that seems kinda easy is to add the refreshing to the EntityProvider, they have access to the DefaultProcessingDatabase so should be easy to expose a new method that calls the appropriate refresh function. However, I don't think EntityProviders should handle refreshes, for me they should care about creating/deleting the entities. So I'd probably not put this behaviour on to the EntityProviders. 

I think refreshing belongs to the catalog. So I am thinking about something like that the CatalogBuilder could return a refresh function and then everyone could decide how they want to use that to refresh their stuff. 

stg like:
```ts
export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  const builder = await CatalogBuilder.create(env);
  builder.addProcessor(new ScaffolderEntitiesProcessor());
  const { processingEngine, router, refresh } = await builder.build();
  
  await processingEngine.start();
  
  refresh(['refreshKey-1', 'refreshKey-2'])

  return router;
}

```

kissmikijr replied:
I actually removed the possibility to trigger refreshes via the RefreshService to be able to tackle it in a separate PR if we decide it that way.

Rugvip reviewed (changes requested):
Feeling we're getting pretty close to :shipit:  😁

Rugvip reviewed (approved):
Nice, thanks! 👍 

Shall we :shipit: as is and leave the triggering of refreshes as a followup?

I think the builder addition could work, especially for use-cases like your own. Looking at #11611 this is very likely to be an extension point though, so I almost doesn't matter how we implement it for now as it'll be migrated to that anyway.

I would like us to explore the EntityProvider approach too though, making it possible for a provider to trigger refreshes through the connection. Or maybe even a new form of abstraction like some form of refresh sources

kissmikijr commented:
> Shall we :shipit: as is and leave the triggering of refreshes as a followup?

I'd say so let's merge this one and I'll create a follow-up issue where we can discuss the different approaches for triggering the refreshes. I can explore both and create some PRs with the different approaches to discuss which fits better.

Rugvip commented:
🎉
